animation_nodes/nodes/mesh/points_scatter.pyx

```diff
@@ -0,0 +1,165 @@
+        triWeights.data[i] = (weights.get(polyIndex1) + weights.get(polyIndex2) + weights.get(polyIndex3)) / weightMax / 3.0
+
+    return polyAmount, triAreas, triWeights
+
+@cython.cdivision(True)
+cdef float triangleArea(Vector3 v1, Vector3 v2, Vector3 v3):
+    cdef Vector3 vs1, vs2, vc
+    vs1.x = v1.x - v3.x
+    vs1.y = v1.y - v3.y
+    vs1.z = v1.z - v3.z
+
+    vs2.x = v2.x - v3.x
+    vs2.y = v2.y - v3.y
+    vs2.z = v2.z - v3.z
+
+    vc.x = vs1.y * vs2.z - vs1.z * vs2.y
+    vc.y = vs1.z * vs2.x - vs1.x * vs2.z
+    vc.z = vs1.x * vs2.y - vs1.y * vs2.x
+    return sqrt(vc.x * vc.x + vc.y * vc.y + vc.z * vc.z) / 2.0
+
+@cython.cdivision(True)
+cdef distributionOfPoints(Py_ssize_t polyAmount, FloatList triAreas, FloatList triWeights,
+                          bint useWeightForDensity, Py_ssize_t seed, Py_ssize_t pointAmount):
+    # Initial distribution of points.
+    cdef double triAreaTotal = 0
+    cdef Py_ssize_t i
+    for i in range(polyAmount):
+        triAreaTotal += triAreas.data[i]
+
+    if triAreaTotal == 0: triAreaTotal = 1
+    triAreaTotal = pointAmount / triAreaTotal
+
+    cdef LongList distribution = LongList(length = polyAmount)
+    cdef Py_ssize_t currentAmount, triPointAmount
+    currentAmount = 0
+    if useWeightForDensity:
+        for i in range(polyAmount):
+            triPointAmount = int(ceil(triAreas.data[i] * triAreaTotal))
+            distribution.data[i] = triPointAmount
+            currentAmount += triPointAmount
+    else:
+        for i in range(polyAmount):
+            triPointAmount = int(ceil(triAreas.data[i] * triAreaTotal * max(triWeights.data[i], 0)))
+            distribution.data[i] = triPointAmount
+            currentAmount += triPointAmount
+
+    # If currentAmount is equal to pointAmount or zero.
+    cdef Py_ssize_t expectedAmount = currentAmount
+    if expectedAmount == pointAmount or currentAmount == 0:
+        return distribution, expectedAmount
+
+    # If currentAmount is not equal to pointAmount.
+    expectedAmount = pointAmount
+    cdef Py_ssize_t step, offsetAmount
+    offsetAmount = abs(pointAmount - currentAmount)
+    step = 1
+    if currentAmount > pointAmount: step = -1
+
+    cdef XoShiRo256StarStar rng = XoShiRo256StarStar(seed)
+    cdef Py_ssize_t j
+    for i in range(offsetAmount):
+        for j in range(polyAmount):
+            index = rng.nextIntWithMax(polyAmount)
+            triPointAmount = distribution.data[index]
+            if triPointAmount == 0: continue
+            distribution.data[index] = triPointAmount + step
+            break
+
+    # Use weight as density.
+    if useWeightForDensity:
+        expectedAmount = 0
+        for i in range(polyAmount):
+            triPointAmount = int(ceil(distribution.data[i] * max(triWeights.data[i], 0)))
+            distribution.data[i] = triPointAmount
+            expectedAmount += triPointAmount
+    return distribution, expectedAmount
```

Comment:
Why can't we join this with the other loop above? It seems to me `useWeightForDensity` grants us an early exist from the function.

Reply:
The reason is, first we make sure each polygon has correct number of vertices so that sum is equal to total points. Then we impose the weight to reduce the number of points per polygon according to their weight.